cli/command/stack/list.go

```diff
@@ -18,20 +22,91 @@ func newListCommand(dockerCli command.Cli) *cobra.Command {
 		Short:   "List stacks",
 		Args:    cli.NoArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			if dockerCli.ClientInfo().HasKubernetes() {
-				kli, err := kubernetes.WrapCli(dockerCli, kubernetes.NewOptions(cmd.Flags()))
-				if err != nil {
-					return err
-				}
-				return kubernetes.RunList(kli, opts)
-			}
-			return swarm.RunList(dockerCli, opts)
+			return runList(cmd, dockerCli, opts)
 		},
 	}
 
 	flags := cmd.Flags()
 	flags.StringVar(&opts.Format, "format", "", "Pretty-print stacks using a Go template")
+	flags.StringSliceVar(&opts.Namespaces, "namespace", []string{}, "Kubernetes namespaces to use")
+	flags.SetAnnotation("namespace", "kubernetes", nil)
+	flags.SetAnnotation("namespace", "experimentalCLI", nil)
 	flags.BoolVarP(&opts.AllNamespaces, "all-namespaces", "", false, "List stacks among all Kubernetes namespaces")
 	flags.SetAnnotation("all-namespaces", "kubernetes", nil)
+	flags.SetAnnotation("all-namespaces", "experimentalCLI", nil)
 	return cmd
 }
+
+func runList(cmd *cobra.Command, dockerCli command.Cli, opts options.List) error {
+	stacks := []*formatter.Stack{}
+	if dockerCli.ClientInfo().HasSwarm() {
+		ss, err := swarm.GetStacks(dockerCli)
+		if err != nil {
+			return err
+		}
+		stacks = append(stacks, ss...)
+	}
+	if dockerCli.ClientInfo().HasKubernetes() {
+		kopts := kubernetes.NewOptions(cmd.Flags())
+		if opts.AllNamespaces || len(opts.Namespaces) == 0 {
+			if dockerCli.ClientInfo().HasAll() {
```

Comment:
`--orchestrator=kubernetes` yields `HasKubernetes()` `true` and `HasAll()` `false`
`--orchestrator=all` yields both `HasKubernetes()` `true` and `HasAll()` `true`
So if `HasAll()` implies `HasKubernetes()`, the reverse is not necessary true.

Reply:
But here `HasAll()` (ie `--orchestrator=all`) is used to enable `--all-namespaces`; I don't see how those relate

Reply:
This is a UX requirement defined in the PRD

> * Provide intuitive mechanisms for switching Kubernetes namespace:
>   * When targeting all orchestrators, target all namespaces by default

Reply:
I'm having trouble conflating these; this may become very confusing:

- configuration has `"orchestrator: all"`: `docker stack ls` shows all stacks (both swarm and k8s)
- I want to view only the `k8s` stacks, so I add `--orchestrator=kubernetes`. Now I no longer see all my stacks, only in the `default` namespace
- confused

Reply:
The thinking behind this was as follows: When you list stacks across orchestrators your intention is to see all the stacks running on your cluster. When you specify `kubernetes` as your orchestrator; you are doing something more specific and want similar behaviour to what `kubectl` gives you.

I can see how this might be confusing for your flow. There will be a _NAMESPACE_ column which should help the user understand what has happened.

Reply:
Ah, see you're referring to the case when a user has set `orchestrator` to `all` in their config file. I don't think many users will do this as they will only be able to list resources. In order to manipulate resources, they will need to specify an orchestrator.